progress/progress.go

```diff
@@ -0,0 +1,100 @@
+package progress
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/term"
+)
+
+type State interface {
+	String() string
+}
+
+type Progress struct {
+	mu  sync.Mutex
+	pos int
+	w   io.Writer
+
+	ticker *time.Ticker
+	states []State
+}
+
+func NewProgress(w io.Writer) *Progress {
+	p := &Progress{w: w}
+	go p.start()
+	return p
+}
+
+func (p *Progress) Stop() bool {
+	for _, state := range p.states {
+		if spinner, ok := state.(*Spinner); ok {
+			spinner.Stop()
+		}
+	}
+
+	if p.ticker != nil {
+		p.ticker.Stop()
+		p.ticker = nil
+		p.render()
+		return true
+	}
+
+	return false
+}
+
+func (p *Progress) StopAndClear() bool {
+	stopped := p.Stop()
+	if stopped {
+		termWidth, _, err := term.GetSize(int(os.Stderr.Fd()))
+		if err != nil {
+			panic(err)
+		}
+
+		// clear the progress bar by:
+		// 1. reset to beginning of line
+		// 2. move up to the first line of the progress bar
+		// 3. fill the terminal width with spaces
+		// 4. reset to beginning of line
+		fmt.Fprintf(p.w, "\r\033[%dA%s\r", p.pos, strings.Repeat(" ", termWidth))
```

Comment:
This is easier to read w/ constants. Also, if p.pos is 0, it will actually move up one space on some terminals which is annoying.

Reply:
Also, you should be able to use "\033[K" to clear to the EOL. This is faster than writing spaces, and shouldn't flicker. You can also use "\033[?25l" to hide the cursor and then "\033[?25h" to show it again which will reduce flicker.

Reply:
I'll use constants once #1133 is merged to avoid conflicts